### PR TITLE
docs(tutorials/blog): add TS 4.9's `satisfies` in combination with `LoaderFunction`

### PR DESCRIPTION
**Pull request by tom-sherman**
Closes: #

- [x] Docs
- [ ] Tests

Testing Strategy:

I don't think this is needed as it's just a utility type. The interesting part happens in userland when we use the `satisfies` keyword.

---

Right now because of the way you need to add types for loaders you can get into a situation where the return type of `useLoaderData` is `never`:

```ts
const loader = () => "not a response";
const data = useLoaderData<typeof loader>();
//    ^? never
```

ideally we'd like typescript to error when we return the string instead of a Response. We can use the new `satisfies` keyword for this:

```ts
// ERROR: Type 'number' is not assignable to type 'Response'
const loader = (() => "not a response") satisfies LoaderFunction;
```

The loader now has a type error but it still allows for the inference of the return type to work with `useLoaderData`.

### CODE REVIEW

File: packages/remix-server-runtime/router/utils.ts [hunk 1]
```diff
@@ -114,6 +114,24 @@ export interface ActionFunction {
...
+ *
+ * ```ts
+ * export const loader = (({ request }) => {
+ *   const url = new URL(request.url);
+ *   return new Response("hello from " + url.pathname);
+ * }) satisfies LoaderFunction;
+ * ```
+ */
+export interface LoaderFunction {
+  (args: LoaderFunctionArgs): Promise<Response> | Response;
```
Comment (tom-sherman) [inline packages/remix-server-runtime/router/utils.ts:132]: We may need to iterate on the return type a little.

Eg. do we want to allow loader functions to return `null | Response`?

File: packages/remix-server-runtime/routeModules.ts [hunk 2]
```diff
@@ -30,6 +30,24 @@ export interface ActionFunction {
...
+ *
+ * ```ts
+ * export const loader = (({ request }) => {
+ *   const url = new URL(request.url);
+ *   return new Response("hello from " + url.pathname);
+ * }) satisfies LoaderFunction;
+ * ```
+ */
+export interface LoaderFunction {
+  (args: DataFunctionArgs): Promise<Response> | Response;
```
Comment (tom-sherman) [inline packages/remix-server-runtime/routeModules.ts:48]: We may need to iterate on the return type a little.

Eg. do we want to allow loader functions to return `null | Response`?

Comment (brophdawg11) [reply packages/remix-server-runtime/routeModules.ts:48]: This already exists later on this this file?

Comment (tom-sherman) [reply packages/remix-server-runtime/routeModules.ts:48]: Oh! I don't have this, maybe because I'm on an experimental remix build (deferred channel)

**Comment by tom-sherman** [comment]
Ok so the Typescript change can be reverted, I got confused because the version of remix I have doesn't have a LoaderFunction type.

@brophdawg11 do you think the docs change is valuable at all?

File: packages/remix-server-runtime/routeModules.ts (hunk 2 above)
Comment (brophdawg11) [reply packages/remix-server-runtime/routeModules.ts:48]: ah that makes sense :)

**Comment by sergiodxa** [comment]
Question, can you use `satisfies` without arrow functions?

**Comment by brophdawg11** [comment]
@tom-sherman actually, since you can technically return anything (`type AppData = any`)  - does satisfies even help in that case?  It might be something where it only helps if you define a more narrowly-typed loader function on your own?

**Comment by tom-sherman** [comment]
@sergiodxa Unfortunately not, it appears the syntax doesn't support it right now  - you can only use `satisfies` with an expression.

@brophdawg11 Oh you're absolutely right 🤦 in which case I don't think this even deserves a place in the docs. It might make sense as a more editorialised blog post I guess (I'm already writing one on my own blog)

**Comment by tom-sherman** [comment]
@sergiodxa A workaround is something like this:

```ts
type Fn = () => number;

function foo() { return "ds" }

foo satisfies Fn; // type error
```

Not recommended tho because compilers will likely output a `foo;` statement.

**Comment by pcattori** [comment]
We're already working on upgrading esbuild to 0.15.13 for `satisfies` support. You can check https://github.com/remix-run/remix/issues/4639 for progress on that.

**Comment by pcattori** [comment]
> @tom-sherman actually, since you can technically return anything (`type AppData = any`) - does satisfies even help in that case? It might be something where it only helps if you define a more narrowly-typed loader function on your own?

There's still value in using `satisfies` even since it will make sure the function params are correct for example.

**Comment by tom-sherman** [comment]
> One `function`-compatible option is:
> 
> ```ts
> (function hello() {
>     return "world"
> }) satisfies () => number;
> ```

@pcattori that's a function expression/anonymous function though. You can't do this for example

```ts
export (function hello() {
    return "world"
}) satisfies () => number;
```

**Comment by pcattori** [comment]
> @pcattori that's a function expression/anonymous function though. You can't do this for example
> 
> ```ts
> export (function hello() {
>     return "world"
> }) satisfies () => number;
> ```

Yea, if you want (1) `export` on the same line as the definition and (2) to use `satisfies`, then I don't think there's a way to do it with a named function.

The only way would be:

```ts
export const hello = (function () {
    return "world"
}) satisfies () => number;
```

but the main benefit of named functions is that they can be referenced before their definition, which you would lose out on with this^.

Personally, I prefer to use arrow functions, but yes its unfortunate for those who use `function` that `satisfies` is not supported better for it AFAIK.

File: packages/remix-server-runtime/routeModules.ts (hunk 2 above)
Comment (pcattori) [reply packages/remix-server-runtime/routeModules.ts:48]: @tom-sherman if you can remove changes to this file, the other doc changes you've made look good to me!

**Comment by tom-sherman** [comment]
Btw, shameless plug 😅 https://tom-sherman.com/blog/remix-typescript-satisfies

**Comment by pcattori** [comment]
> https://tom-sherman.com/blog/remix-typescript-satisfies

FYI we export a `LoaderFunction` type from `@remix-run/node` (and the other runtime packages too), so you can use that instead of defining `Loader` yourself: `satisfies LoaderFunction`

**Review by pcattori — APPROVED** [review]

**Comment by pcattori** [comment]
Looks like tests are failing because eslint can't parse Typescript 4.9 `satisfies` correctly. Probably need to upgrade esbuild or the typescript plugin or something like that.

File: docs/tutorials/blog.md [hunk 3]
```diff
@@ -192,6 +192,33 @@ export default function Posts() {
 
 Hey, that's pretty cool. We get a pretty solid degree of type safety even over a network request because it's all defined in the same file. Unless the network blows up while Remix fetches the data, you've got type safety in this component and its API (remember, the component is already its own API route).
 
+We can take the type safety a step further in TypeScript 4.9.0 using the `satisfies` keyword. This will ensure that we always return a type of `Response` from the loader instead of accidentally returning another type which would result in a runtime error.
+
+```ts filename=app/routes/posts/index.tsx lines=[1,22]
+import { json, LoaderFunction } from "@remix-run/node";
```
Comment (MichaelDeBoey) [inline docs/tutorials/blog.md:198]: We put types on a separate line in our docs/examples

```suggestion
import type { LoaderFunction } from "@remix-run/node";
import { json } from "@remix-run/node";
```

File: docs/tutorials/blog.md [hunk 4]
```diff
@@ -192,6 +192,33 @@ export default function Posts() {
 
 Hey, that's pretty cool. We get a pretty solid degree of type safety even over a network request because it's all defined in the same file. Unless the network blows up while Remix fetches the data, you've got type safety in this component and its API (remember, the component is already its own API route).
 
+We can take the type safety a step further in TypeScript 4.9.0 using the `satisfies` keyword. This will ensure that we always return a type of `Response` from the loader instead of accidentally returning another type which would result in a runtime error.
+
+```ts filename=app/routes/posts/index.tsx lines=[1,22]
```
Comment (MichaelDeBoey) [inline docs/tutorials/blog.md:197]:
```suggestion
```ts filename=app/routes/posts/index.tsx lines=[1,23]
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Comment by tom-sherman** [comment]
> I guess you already knew that since this PR uses LoaderFunction in the docs 😅

@pcattori Weird one, I wasn't actually aware of that type when I wrote the blob because I'm on an experimental channel (deffered) that doesn't have that type 😆

File: docs/tutorials/blog.md [hunk 5]
```diff
@@ -195,7 +195,8 @@ Hey, that's pretty cool. We get a pretty solid degree of type safety even over a
 We can take the type safety a step further in TypeScript 4.9.0 using the `satisfies` keyword. This will ensure that we always return a type of `Response` from the loader instead of accidentally returning another type which would result in a runtime error.
 
 ```ts filename=app/routes/posts/index.tsx lines=[1,22]
```
Comment (machour) [inline docs/tutorials/blog.md]:
```suggestion
```ts filename=app/routes/posts/index.tsx lines=[1,23]
```

Comment (machour) [reply docs/tutorials/blog.md]: Weird, I was seeing an outdated version

**Comment by moishinetzer** [comment]
Is there any updated best practice/solution to solve this?

I want to be able to take advantage of `useLoaderData<typeof loader>()` AND make sure that my loader functions return the correct types.

**Comment by tom-sherman** [comment]
@moishinetzer The simplest way to do that would be to add a return type annotation to the loader eg.

```ts
export async function loader({ context }: LoaderArgs): Promise<TypedResponse<{ posts: Post[], tags: string[] }>> {
  const blog = new BlogData(context);

  const [posts, tags] = await Promise.all([
    blog.listAllPosts(),
    blog.listAllTags(),
  ]);

  return json({
    posts,
    tags,
  });
}

const { posts, tags } = useLoaderData<typeof loader>();
```

You can use `satisfies` if you wanted to assert a partial data shape at the loader level and have the concrete return type inferred.

**Review by machour — CHANGES_REQUESTED** [review]
@tom-sherman Could you rebase this to target `main`?
Documentation & templates updates go directly to main

File: docs/tutorials/blog.md [hunk 6]
```diff
@@ -195,6 +195,34 @@ export default function Posts() {
 
 Hey, that's pretty cool. We get a pretty solid degree of type safety even over a network request because it's all defined in the same file. Unless the network blows up while Remix fetches the data, you've got type safety in this component and its API (remember, the component is already its own API route).
 
+We can take the type safety a step further in TypeScript 4.9.0 using the `satisfies` keyword. This will ensure that we always return a type of `Response` from the loader instead of accidentally returning another type which would result in a runtime error.
+
+```ts filename=app/routes/posts/index.tsx lines=[1,23]
```
Comment (MichaelDeBoey) [inline docs/tutorials/blog.md:200]:
```suggestion
```ts filename=app/routes/posts/index.tsx
```

File: docs/tutorials/blog.md [hunk 7]
```diff
@@ -195,6 +195,34 @@ export default function Posts() {
...
+import type { LoaderFunction } from "@remix-run/node";
+import { json } from "@remix-run/node";
```
Comment (MichaelDeBoey) [inline docs/tutorials/blog.md:202]:
```suggestion
import type { LoaderFunction } from "@remix-run/node"; // or cloudflare/deno
import { json } from "@remix-run/node"; // or cloudflare/deno
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Comment by pcattori** [comment]
Looks like the upstream issues with eslint plugins and TS 4.9 have been resolved as `yarn lint` passes locally for me with the suggested edits.

That said, we are no longer using this tutorial in the v2 docs, so closing this PR.

**Pull request closed without merging**